### PR TITLE
[AMDGPU][True16][MC] 16bit vsrc and vdst support in MC

### PR DESCRIPTION
**Pull request by broxigarchen**
This is a large patch includes the MC level support for V_CVT_F16_F32, V_CVT_F32_F16 and V_LDEXP_F16 in true16 format.

This patch includes the asm/disasm changes to encode/decode the 16bit vsrc, vdst and src modifieres for vop and dpp format. This patch is a dependency for many 16 bit instructions while only three instructions are updated to make it easier to review.

There will be another patch to support these three instructions in the codeGen level, this patch just replaces these two instructions with its fake16 format.

### CODE REVIEW

**Comment by broxigarchen** [comment]
Hi @arsenm @Sisyph @kosarev This PR is ready for review. Thanks!

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_from_vop1.s
Comment (Sisyph) [inline llvm/test/MC/AMDGPU/gfx11_asm_vop3_from_vop1.s:2]: Should there be a test file for fake16 gfx11_asm_vop3_from_vop1-fake16.s ?

File: llvm/test/MC/AMDGPU/gfx12_asm_vop1.s
Comment (Sisyph) [inline llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:602]: It looks like there should be mc tests for gfx12 dpp8, dpp16, vop3, vop3 dpp8, vop3dpp16. And it looks like we need the fake16 files for those

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_from_vop1.txt
Comment (Sisyph) [inline llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_from_vop1.txt:574]: Add GFX12 disassembler tests?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/inst-select-fptosi.mir
Comment (broxigarchen) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/inst-select-fptosi.mir:21]: Opened a PR here https://github.com/llvm/llvm-project/pull/106073

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_from_vop1.s
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_from_vop1.s:2]: opened two PR for adding missing test files
gfx11 https://github.com/llvm/llvm-project/pull/106089
gfx12 https://github.com/llvm/llvm-project/pull/106093

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_from_vop1.txt
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_from_vop1.txt:574]: The gfx12 dasm test need to be updated first before updating the change for v_cvt_fxx. Opened a PR here https://github.com/llvm/llvm-project/pull/106469

File: llvm/lib/Target/AMDGPU/SIInstructions.td
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/SIInstructions.td]: done

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.td]: done

File: llvm/lib/Target/AMDGPU/VOP1Instructions.td
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/VOP1Instructions.td]: done

File: llvm/test/CodeGen/AMDGPU/fix-sgpr-copies-f16-true16.mir
Comment (broxigarchen) [reply llvm/test/CodeGen/AMDGPU/fix-sgpr-copies-f16-true16.mir:3]: changed

File: llvm/test/CodeGen/AMDGPU/fix-sgpr-copies-f16.mir
Comment (broxigarchen) [reply llvm/test/CodeGen/AMDGPU/fix-sgpr-copies-f16.mir:37]: The fake16 flow is working. The reason I marked it as XFAIL is because the name of V_CVT_F16_U16_fake16 is currently  V_CVT_F16_U16_t16 (t16 but use fake16 profile).

I can use V_CVT_F16_U16_t16  in the mir file to get this test pass for now and I can update the V_CVT_F16_U16 later

File: llvm/test/MC/AMDGPU/gfx12_asm_vop1.s
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop1.s:602]: added

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_from_vop1.txt
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_from_vop1.txt:574]: added

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_from_vop1.txt
Comment (broxigarchen) [inline llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_from_vop1.txt:512]: one thing I noticed is that in the dasm test, the disassemble of "vop3_from_vop1" instructions do not error out when hi16 bit is set in fake16 flow. For now the disassembler just ignore the hi16 bit and use lo16.

In gfx11, we have all hi16 in a seperated test file vop3_from_vop1_hi.txt while do not test the fake16 on it. So I did not notice this before.

@Sisyph @kosarev Do we need to error out in disassebler in this case?

**Comment by broxigarchen** [comment]
ping! Hi all reviewers. Otjher than a few small issues, this PR is ready for review.

File: llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_from_vop1.txt
Comment (kosarev) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_from_vop1.txt:512]: So far just leaving the fake16 support as it is was enough, so that should be fine.

Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx12_dasm_vop3_from_vop1.txt:512]: Cool Thanks!

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_from_vop1.txt
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_from_vop1.txt:578]: added

File: llvm/test/MC/AMDGPU/gfx12_asm_vop1_t16_promote.s
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop1_t16_promote.s]: done

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp8_from_vop1.s
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp8_from_vop1.s:146]: added

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16_from_vop1.s
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_dpp16_from_vop1.s:518]: added

File: llvm/test/MC/AMDGPU/gfx11_asm_vop3_from_vop1.s
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop3_from_vop1.s:542]: added

File: llvm/test/MC/AMDGPU/gfx12_asm_vop3_from_vop1_dpp16.s
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx12_asm_vop3_from_vop1_dpp16.s:518]: added

File: llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_dpp8_from_vop1.txt
Comment (broxigarchen) [reply llvm/test/MC/Disassembler/AMDGPU/gfx11_dasm_vop3_dpp8_from_vop1.txt:151]: added

**Review by Sisyph — COMMENTED** [review]
LGTM. Thanks for this significant progress. Please wait for reviews from others who are not as close to this code. @arsenm @jayfoad @rampitec

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td
Comment (kosarev) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.td:1256]: Nit: we don't really use or need that default value, do we? +Same for other similar classes below.

File: llvm/test/MC/AMDGPU/gfx11_asm_vop1.s
Comment (kosarev) [inline llvm/test/MC/AMDGPU/gfx11_asm_vop1.s:406]: Just to make sure we are on the same page with this: these patches make no changes to the MC tests, such as reordering of instructions, that would complicate merging them in downstream, right?

**Review by kosarev — APPROVED** [review]
LGTM with a nit.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td
Comment (broxigarchen) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.td:1256]: Yes we don't really need to have a default value. But it does not harm anything though. Should I remove it?

File: llvm/test/MC/AMDGPU/gfx11_asm_vop1.s
Comment (broxigarchen) [reply llvm/test/MC/AMDGPU/gfx11_asm_vop1.s:406]: Since this patch is big, I might do cherry-pick myself so it won't be too complicate for the downstream merging.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td
Comment (kosarev) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.td:1256]: This looks a bit like an invitation to use the default value, which I'm not sure would be a good idea for these particular classes, considering the default semantics is not obvious in at least one of the mainline or the downstream branches, if not both. Also, adding things that we don't really use goes a bit against common practice for reasons such as wasting resources changing and maintaining them. But probably not a big deal either way, so up to you to decide.

**Comment by broxigarchen** [comment]
> LGTM. Thanks for this significant progress. Please wait for reviews from others who are not as close to this code. @arsenm @jayfoad @rampitec

Quick ping! Thanks!

**Comment by jayfoad** [comment]
I did not review everything in detail, but it looks OK to me.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.td:1788]: Nit: might be simpler to swap this condition to check `!eq(VT.Value, f32.Value)` since there is only one 32-bit fp type to check against. But I see you are just copying an existing pattern.

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIInstrInfo.td
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.td:1788]: > Do you mean we can change it from `!if (!or(!eq(VT.Value, f16.Value), !eq(VT.Value, bf16.Value)), A, B)` `!if (!eq(VT.Value, f32.Value), B, A)`

Yes exactly.

**Pull request merged**
